gtda/utils/validation.py

```diff
@@ -204,45 +220,90 @@ def check_point_clouds(X, distance_matrix=False, **kwargs):
     -------
     Xnew : ndarray or list
         The converted and validated object.
 
     """
-    kwargs_ = {'force_all_finite': not distance_matrix}
+    kwargs_ = {'force_all_finite': not distance_matrices}
     kwargs_.update(kwargs)
-    if hasattr(X, 'shape'):
+    kwargs_.pop('allow_nd', None)
+    kwargs_.pop('ensure_2d', None)
+    if hasattr(X, 'shape') and hasattr(X, 'ndim'):
         if X.ndim != 3:
-            raise ValueError("ndarray input must be 3D.")
-        return check_array(X, allow_nd=True, **kwargs_)
+            if X.ndim == 2:
+                extra_2D = \
+                    "\nReshape your input X using X.reshape(1, *X.shape) or " \
+                    "X[None, :, :] if X is a single point cloud/distance " \
+                    "matrix/adjacency matrix of a weighted graph."
+            else:
+                extra_2D = ""
+            raise ValueError(
+                f"Input must be a single 3D array or a list of 2D arrays. "
+                f"Array of dimension {X.ndim} passed." + extra_2D)
+        if (X.shape[1] != X.shape[2]) and distance_matrices:
+            raise ValueError(
+                f"Input array X must have X.shape[1] == X.shape[2]: "
+                f"{X.shape[1]} != {X.shape[2]} passed.")
+        elif (X.shape[1] == X.shape[2]) and not distance_matrices:
+            warn(
+                "Input array X has X.shape[1] == X.shape[2]. This is "
+                "consistent with a collection of distance/adjacency "
+                "matrices, but the input is being treated as a collection "
+                "of vectors in Euclidean space.",
+                DataDimensionalityWarning, stacklevel=2)
+        Xnew = _check_array_mod(X, **kwargs_, allow_nd=True)
     else:
-        if not distance_matrix:
-            reference = X[0].shape[1]  # Embedding dimension of first sample
-            if not reduce(
-                    and_, (x.shape[1] == reference for x in X[1:]), True):
-                warn("Not all point clouds have the same embedding dimension.")
-
-    has_check_failed = False
-    messages = []
-    Xnew = []
-    for i, x in enumerate(X):
-        try:
-            Xnew.append(check_array(x, **kwargs_))
-            messages = ['']
-        except ValueError as e:
-            has_check_failed = True
-            messages.append(str(e))
-    if has_check_failed:
-        raise ValueError("The following errors were raised by the inputs: \n"
-                         "\n".join(messages))
+        has_check_failed = False
+        messages = []
+        Xnew = []
+        for i, x in enumerate(X):
+            try:
+                xnew = _check_array_mod(x, **kwargs_, ensure_2d=True)
+                if distance_matrices:
+                    if not x.shape[0] == x.shape[1]:
+                        raise ValueError(
+                            f"All arrays must be square: {x.shape[0]} rows "
+                            f"and {x.shape[1]} columns found in this array.")
```

Comment:
What do you think about including this in `_check_array_mod`, since `is_square` is a property of an individual array?

Reply:
I would prefer it here as I didn't intend for `_check_array_mod` to add new parameters. Also: 1) the error message would have to be changed; 2) this would not cover the 3D case.